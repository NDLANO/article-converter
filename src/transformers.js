/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { replaceEmbedsInHtml, addClassToTag, replaceStartAndEndTag } from './replacer';
import { getEmbedsFromHtml } from './parser';
import { extractCopyrightInfoFromEmbeds } from './extractCopyrightInfo';
import plugins from './plugins';


// Changes aside tags to accommodate frontend styling
export const asideReplacers = [
  replaceStartAndEndTag('aside', '<aside><div class="c-aside__content">', '</div><button class="c-button c-aside__button"></button></aside>'),
  addClassToTag('aside', 'c-aside u-1/3@desktop'),
];

export async function transformContentAndExtractCopyrightInfo(content, lang, accessToken) {
  const embeds = await getEmbedsFromHtml(content);
  const embedsWithResources = await Promise.all(embeds.map((embed) => {
    const plugin = plugins.find(p => p.resource === embed.resource);
    if (plugin && plugin.fetchResource) {
      return plugin.fetchResource(embed, accessToken);
    }
    return embed;
  }));

  const contentWithReplacedEmbeds = await replaceEmbedsInHtml(embedsWithResources, lang)(content);

  return {
    html: asideReplacers.reduce((html, f) => f(html), contentWithReplacedEmbeds),
    copyrights: extractCopyrightInfoFromEmbeds(embedsWithResources),
  };
}
