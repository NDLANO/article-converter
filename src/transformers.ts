/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { performance } from 'perf_hooks';
import { CheerioAPI } from 'cheerio';
import { compact } from 'lodash';
import { replaceEmbedsInHtml } from './replacer';
import { getEmbedsFromHtml } from './parser';
import getEmbedMetaData from './getEmbedMetaData';
import createPlugins from './plugins';
import log from './utils/logger';
import { htmlTransforms } from './htmlTransformers';
import {
  PluginUnion,
  EmbedType,
  LocaleType,
  TransformOptions,
  ResponseHeaders,
} from './interfaces';
import { findPlugin } from './utils/findPlugin';
import { mergeResponseHeaders } from './utils/mergeResponseHeaders';

function logIfLongTime(start: number, timeout: number, action: string, obj: any) {
  const elapsedTime = performance.now() - start;
  if (elapsedTime > timeout) {
    console.warn(
      `Action '${action}' took ${elapsedTime.toFixed(
        2,
      )}ms, longer than time to warn of ${timeout}ms...`,
      obj,
    );
  }
}

async function executeHtmlTransforms(
  content: CheerioAPI,
  lang: LocaleType,
  options: TransformOptions,
): Promise<void> {
  // We use this reduce trick so the transformations happen in sequence even if the function is asynchronous
  const starterPromise = Promise.resolve();
  await htmlTransforms.reduce(
    (
      maybePromise: Promise<void>,
      replacer: (content: CheerioAPI, lang: LocaleType, options: TransformOptions) => void,
    ) => maybePromise.then(() => replacer(content, lang, options)),
    starterPromise,
  );
}

export async function getEmbedsResources(
  embeds: EmbedType[],
  accessToken: string,
  feideToken: string,
  lang: LocaleType,
  plugins: PluginUnion[],
): Promise<EmbedType[]> {
  return Promise.all(
    embeds.map(async (embed) => {
      const plugin = findPlugin(plugins, embed);
      if (plugin && plugin.fetchResource) {
        const startStamp = performance.now();
        try {
          const resource = await plugin.fetchResource(embed, accessToken, lang, feideToken);
          logIfLongTime(startStamp, 500, `Fetching resource`, embed.data);
          return resource;
        } catch (e) {
          log.warn('Failed to fetch embed resource data for ', embed.data);
          log.warn(e);
          logIfLongTime(startStamp, 500, `Failed fetching resource`, embed.data);
          return {
            ...embed,
            status: 'error',
          };
        }
      }
      return embed;
    }),
  );
}

export type TransformFunction = (
  content: CheerioAPI,
  headers: Record<string, string>,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  visualElement: { visualElement: string } | undefined,
  options: TransformOptions,
) => Promise<{ html: string | null; embedMetaData: any; responseHeaders?: ResponseHeaders }>;

export const transform: TransformFunction = async (
  content,
  headers,
  lang,
  accessToken,
  feideToken,
  visualElement,
  options,
) => {
  if (visualElement?.visualElement && options?.showVisualElement) {
    content('body').prepend(`<section>${visualElement.visualElement}</section>`);
  }

  const transformOptions = { transform, ...options };
  const plugins = createPlugins(transformOptions);
  const embeds = await getEmbedsFromHtml(content);
  const embedsWithResources = await getEmbedsResources(
    embeds,
    accessToken,
    feideToken,
    lang,
    plugins,
  );

  const htmlHeaders = await replaceEmbedsInHtml(embedsWithResources, lang, plugins);
  const embedMetaData = await getEmbedMetaData(embedsWithResources, lang, plugins);

  const fetchedResourceHeaders = compact(embedsWithResources.map((x) => x.responseHeaders));

  const allResponseHeaders = [...fetchedResourceHeaders, ...htmlHeaders, headers];
  const responseHeaders = mergeResponseHeaders(allResponseHeaders);
  await executeHtmlTransforms(content, lang, options);

  return {
    html: content('body').html(),
    embedMetaData,
    responseHeaders,
  };
};
