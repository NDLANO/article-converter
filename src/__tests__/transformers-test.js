/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { asideReplacers } from '../transformers';


test('transformers/asideReplacers changes aside to accommodate frontend styling', () => {
  const content = `
  <section>
    <p>Lorem ipsum dolor sit amet...</p>
    <aside><h2>Test1</h2><div>Stuff</div></aside>
  </section>
    <aside><h2>Test2</h2><div>Other stuff</div></aside>
    <p>Lorem ipsum dolor sit amet...</p>
  </section>`;

  const result = asideReplacers.reduce((html, f) => f(html), content);

  expect(result).toMatch('<aside class="c-aside u-1/3@desktop"><div class="c-aside__content"><h2>Test1</h2><div>Stuff</div></div><button class="c-button c-aside__button"></button></aside>');
  expect(result).toMatch('<aside class="c-aside u-1/3@desktop"><div class="c-aside__content"><h2>Test2</h2><div>Other stuff</div></div><button class="c-button c-aside__button"></button></aside>');
});
