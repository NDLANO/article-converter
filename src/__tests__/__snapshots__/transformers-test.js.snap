// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformers/htmlTransforms adds display block to math tags 1`] = `
"<html><head></head><body><section>
    <math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\">
      <mi>f</mi><mo>&apos;</mo>
      <mfenced><mrow><mn>0</mn><mo>,</mo><mn>5</mn></mrow></mfenced>
      <mo>=</mo><mn>2</mn><mo>&#xB7;</mo><mn>0</mn><mo>,</mo><mn>5</mn><mo>=</mo><mn>1</mn>
    </math>
    <math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\">
      <mn>0</mn><mo>,</mo><mn>5</mn>
    </math>
  </section>
  </body></html>"
`;

exports[`transformers/moveReactPortals 1`] = `
"<html><head></head><body><section>
    <figure>
      <p>Lorem ipsum dolor sit amet...</p>
      
    </figure>
  </section><div data-react-universal-portal=\\"true\\"><h2>Modal dialog</h2><div>Stuff</div></div></body></html>"
`;
